### DISCUSSION TITLE
How to run llama2 in a fast way?

### DISCUSSION THREAD

**Discussion by RuoshanLan** [oobabooga/text-generation-webui#4710] [Q&A] [answered]
Hello, I am trying to run llama2-70b-hf with 2 Nvidia A100 80G on Google cloud. 
The command I am using is to load model is:  `python [server.py]--public-api --share --model meta-llama_Llama-2-70b-hf --auto-devices --gpu-memory 79 79`

However, I found that the model runs slow when generating. The speed is only about 7 tokens/s.
To speed up, I add --load-in-4bit/--load-in-8bit when loading model, as follows
`python [server.py]--public-api --share --model meta-llama_Llama-2-70b-hf --auto-devices --gpu-memory 79 79 --load-in-8bit`
But then the speed becomes even much slower, as you can see from this picture:

![image](https://github.com/oobabooga/text-generation-webui/assets/10786204/7ffdd4b2-d5f8-45a5-b3d8-6c00108857c2)

My question is whether this speed is normal. Is there any method to speed up generation maybe by using some commands when loading?

Thank you!

**Comment by HiroseKoichi** [comment, marked as answer]
It looks like you're running an unquantized 70B model using transformers. If you want more speed, then you'll need to run a quantized version of it, such as GPTQ or GGUF. Since you have access to 160GB of VRAM, I recommend running GGUF quantization at Q_8 or Q_6 and offloading all layers to the GPU. Q_8 is essentially identical to FP-16 but with significantly lower VRAM requirements, and Q_6 has a slight dip in quality in exchange for lower VRAM requirements. GGUF utilizes the llama.cpp loader. This will not only be much faster, but you can also use a much larger context size as well.

[Here](https://huggingface.co/TheBloke/Llama-2-70B-GGUF) is a link to the GGUF quantization of LLama-2-70B, but I would recommend using a fine-tuned 70B instead of standard LLama-2. You can find a large list of 70B GGUF quantizations [here](https://huggingface.co/models?search=TheBloke/70B%20GGUF), which is done by [TheBloke](https://huggingface.co/TheBloke).

**Reply by RuoshanLan** [reply]
Hey, @HiroseKoichi thank you so much for your detailed explanation! Just to confirm I understand it correctly, here the command I will use: python [server.py]--public-api --share --model llama-2-70b.Q8_0.gguf --loader llama.cpp ---n-gpu-layers 128 --tensor_split 79,79

But I still have another question, I check the link to the GGUF quantization of LLama-2-70B, it says llama-2-70b.Q8_0.gguf has **extremely low quality loss.** Does that mean the performance will still be slightly reduced compared with original one (without quantization)? If so, do you know whether there is a way to do acceleration without affecting performance? e.g., maybe I can use more GPU for it?

**Reply by HiroseKoichi** [reply]
> Hey, @HiroseKoichi thank you so much for your detailed explanation! Just to confirm I understand it correctly, here the command I will use: python [server.py]--public-api --share --model llama-2-70b.Q8_0.gguf --loader llama.cpp ---n-gpu-layers 128 --tensor_split 79,79

Yes, that is correct.

> But I still have another question, I check the link to the GGUF quantization of LLama-2-70B, it says llama-2-70b.Q8_0.gguf has **extremely low quality loss.** Does that mean the performance will still be slightly reduced compared with original one (without quantization)? If so, do you know whether there is a way to do acceleration without affecting performance? e.g., maybe I can use more GPU for it?

This is technically true since there is an observable difference on synthetic benchmarks, but the difference between Q_8 and FP-16 is so small that they're functionally identical for inference, but Q_8 has half the file size and VRAM requirements. If I remember correctly, the difference between Q_6 and FP-16 was somewhere around 0.1% or less. 

You can find a bit more information [here](https://github.com/ggerganov/llama.cpp/pull/1684) and [here](https://github.com/ggerganov/llama.cpp#quantization), but realistically, you won't find any difference in quality with Q_8, and you would have to look pretty hard to find a difference with Q_6.

**Reply by RuoshanLan** [reply]
@HiroseKoichi Thank you very much! I tested the gguf version and it works much faster(13token/s) as expected. But I observed one weird phenomenon when I run this llama2-70b-hf model as a transformer. 

This thing is when I run the model with a command like following:
`python [server.py]--public-api --share --model meta-llama_Llama-2-70b-hf --auto-devices --gpu-memory 79 79`. The speed is 7 token/s
But when I run it with additional `--load-in-8bit` or `--load-in-4bit`. It become even slower than the original one.

According to the [wiki](https://github.com/oobabooga/text-generation-webui/wiki/04-%E2%80%90-Model-Tab), if I don't use 8bit, 4bit or bf16, It will use float16 to load model, which should be slowest among all options. But why here it is faster than "8bit" and "4bit"? Do you have some clues? Thank you again for your help!

**Reply by HiroseKoichi** [reply]
From the [wiki](https://github.com/oobabooga/text-generation-webui/wiki/04-%E2%80%90-Model-Tab):
> load-in-8bit: Load the model in 8-bit precision using bitsandbytes. The 8-bit kernel in that library has been optimized for training and not inference, so load-in-8bit is slower than load-in-4bit (but more accurate).

I have not done any training/fine-tuning myself, so I could be wrong, but I'll try to answer using what I can gather from the wiki page and the [bitsandbytes](https://github.com/TimDettmers/bitsandbytes) github page.

When fine-tuning a model, you always use the full FP-16 model, but this can take a lot of VRAM, so --load-in-8bit and --load-in-4bit reduce the VRAM requirements by using bitsandbytes to reduce the precision, but this extra computation also reduces the speed. This is only useful for training and not inference, so unless you're wanting to fine-tune a model, then --load-in-8bit and --load-in-4bit will not benefit you in any way, and you should stick to quantized models if you want increased inference speed.
